### ISSUE TITLE
Virtua Chess does not work

### ISSUE THREAD

**Issue by Vort** [bochs-emu/Bochs#412]
After Virtua Chess demo is started in MS-DOS, General Protection Fault error appears.
This may be not the only problem with this game, I suspect it makes unusual manipulations with VGA hardware.
I tested it on real hardware and confirm that it works there.

![bochs_vchess](https://github.com/user-attachments/assets/ef4bd2f3-c4c0-44dc-9912-6b31050d872f)

Test files: [VChess.zip](https://github.com/user-attachments/files/18125282/VChess.zip)
Version: 0e420c0

**Comment by vruppert** [comment]
I can confirm that this game has issues on Bochs, Dosbox and Qemu. Qemu hangs after reporting cpu and speed and with Dosbox the game stops after reporting "FATAL: card not 100% VGA compatible". We need to find out what should happen before this exception.

**Comment by fysnet** [comment]
I think it has to do with Executable selectors.

To show this may be true, start the debugger, and set a linear break at 0x1002F3E6
`lb 0x1002F3E6`
Then `c` (continue) until it breaks.
You will be at a `call .-30807 (0x10027b94)`. Step once to a `pusha`.
Now step through until you get to the `mov al,byte ptr ds:[esi]` instruction. Don't execute it yet. View the GDT. The DS selector is `0x33` (0x30 with PL = 3).
Now step once, executing the `mov al,byte ptr ds:[esi]` instruction and there is where the GPF handler takes over.

The debugger needs a little work since the `limit` column for this and two other selectors is blank. Also, it shows `32-bit code` and `32-bit data` for these. It should show `executable`?

Anyway, I have nowhere near the knowledge that Stanislav obviously has on this subject, but shouldn't a read access to an executable segment still be valid? I am asking, not telling :-)

I may be all wrong here, though I am simply pointing out where the GPF happens, hoping that this might help further diagnose the issue.

Thanks.

**Comment by fysnet** [comment]
One of the errors is at line 1423
https://github.com/bochs-emu/Bochs/blob/0e420c096aa08b68b43aa7e8a313003d691b1c71/bochs/gui/enh_dbg.cc#L1420-L1426
and then at line 1501
https://github.com/bochs-emu/Bochs/blob/0e420c096aa08b68b43aa7e8a313003d691b1c71/bochs/gui/enh_dbg.cc#L1501

Line 1501, with an example base address of `0x10000000` will produce a 11-byte null-terminated string, with byte 10 being the null. Since line 1423 only allows for 10 bytes (40 - 30), when we write `0x10000000` to `cols[1]`, it actually places a null byte in the first byte of `cols[2]`.

Making a modification of:
```
-    char gdttxt[90];
+    char gdttxt[100];
```
and
```
-    cols[2]= gdttxt + 40;
-    cols[3]= gdttxt + 80;
+    cols[2]= gdttxt + 50;
+    cols[3]= gdttxt + 90;
```
will fix this.

**Comment by fysnet** [comment]
fix added to #411

**Comment by Vort** [comment]
> Now step through until you get to the mov al,byte ptr ds:[esi] instruction.

The thing is `esi` is `0` at this point, so this is basically null pointer dereference.
It is needed to understand why it is zero.

**Comment by Vort** [comment]
I think crash happens because of failed VGA check in function `lb 0x1002EAFB`.
`es` have base of `0xA0000`, so `es:0x1A000` points to `0xBA000`.
Code expects that after writing `0xA0`, subseqent read will return `0x51`, but with Bochs, result stays at `0xA0`.
![image](https://github.com/user-attachments/assets/51a16bd5-e9b6-4c1a-8958-502155c6d46e)

**Comment by fysnet** [comment]
On a side note, the GDT entry that ES points to after the instruction at 0x1002EB4F is an entry in error. It has both 'L' and 'D/B' bits set.

I think the issue here is the `BX_VGA_THIS s.sequencer.odd_even_dis` flag. I added a check to see if the `odd_even` was enabled before checking if the odd_even_dis flag was used and the above test passed. 

```
--- C:\bochs\bochs\iodev\display\vgacore_.cc	2024-12-14 17:59:50.655319800 -0700
+++ C:\bochs\bochs\iodev\display\vgacore.cc	2024-12-14 17:57:31.501960000 -0700
@@ -1049,6 +1049,8 @@
           if (BX_VGA_THIS s.graphics_ctrl.odd_even)
             BX_DEBUG(("io write: 0x3cf: mode reg: value = 0x%02x",
               (unsigned) value));
+          // I don't know if this addition is actually necessary!
+          else
+            BX_VGA_THIS s.sequencer.odd_even_dis = 0;
           if (BX_VGA_THIS s.graphics_ctrl.shift_reg)
             BX_DEBUG(("io write: 0x3cf: mode reg: value = 0x%02x",
               (unsigned) value));
@@ -1737,7 +1739,7 @@
   if (BX_VGA_THIS s.sequencer.chain_four) {
     // Mode 13h: 320 x 200 256 color mode: chained pixel representation
     return BX_VGA_THIS s.memory[offset];
-  } else if (!BX_VGA_THIS s.sequencer.odd_even_dis) {
+  } else if (BX_VGA_THIS s.graphics_ctrl.odd_even && !BX_VGA_THIS s.sequencer.odd_even_dis) {
     Bit8u plane = (read_map_select & 2) | (offset & 1);
     return BX_VGA_THIS s.memory[((offset & ~1) << 2) | plane];
   }
@@ -1868,7 +1870,7 @@
       }
     }
     return;
-  } else if (!BX_VGA_THIS s.sequencer.odd_even_dis) {
+  } else if (BX_VGA_THIS s.graphics_ctrl.odd_even && !BX_VGA_THIS s.sequencer.odd_even_dis) {
     Bit8u plane = offset & 1;
     Bit8u mask = sequ_map_mask & (0x05 << plane);
     if (mask > 0) {
```

However, a little further and it got to a
```
  mov  ax,0x4F00
  int  10h
```
and hung.

The GPF is gone, but it never changes to the video screen mode. This could be unrelated, it could be related, I don't know, but it may be something to look at.

**Comment by vruppert** [comment]
The VGA core did not handle the read and write modes (including data rotation and raster operations) in odd/even mode. I have now applied a set of changes to fix this (TODO: read mode 1 - not used here).

**Comment by fysnet** [comment]
I can confirm that fixed it. Thank you.

However, shortly later the simulation seems to freeze again.

Start the simulation with the debugger and `lb 0x1002eb84`, then `continue`.
This will get to the point to verify that the fix above is valid.
Then `lb 0x12569`, `continue`.
This will get to just before the:
```
   mov  di,0x08D0
   mov  ax,0x4F00
   int  0x10
```
This will get the video information.
Place a break just after the `int 0x10` at the `cmp  ax,0x4F00`, and `continue`.
Do a physical dump of the returned information at `0x108B0`.
Notice that the list of video screen numbers goes to 0x192 and then the 0xFFFF marker, for a total of 50 entries.
Place a break at `0x1258B` and `continue` multiple times watching the value in AX, stopping when AX = 0x190.
Now hit `continue` one more time to see that AX is not 0x192 as expected but back to 0x100.

Look closely at the code.
The code is reading the entries, starting with the first, from 0x0FFE:08F2 (es:di) and placing them at 0x0FFE:0950 (ds:si) allowing space for only 47 entries.  This version of MS-DOS is assuming that there will not be more than 47 entries...

To see that this is the case, start over following the directions above and instead of hitting `continue` at screen mode 0x190, change AX to 0xFFFF and hit continue to terminate the loop and continue on.

The game will continue on, except that there is no mouse driver installed.

The current VGA BIOS has too many Video Entries for this version of MS-DOS.  :-)

**Comment by fysnet** [comment]
P.S. The VGA code may not be quite correct yet. The rendered text in the video mode chosen (by default?) isn't very clear.

Thanks for the fix to the initial issue given.

**Comment by Vort** [comment]
> Edit: The VGABIOS-elpin-2.40 BIOS does not suffer this issue and goes right to the 'No mouse driver' screen.

Do you use latest VGABIOS binaries?
No more hangs for me.

> P.S. The VGA code may not be quite correct yet. The rendered text in the video mode chosen (by default?) isn't very clear.

Yes, something is not right here:
![image](https://github.com/user-attachments/assets/2948ed36-4c1e-4b18-be47-bbaf81a6f72b)

---

I know about one more problem with this game:
When I launch it from 800 MB partition, it just shows `16 bit code is too large` and exits.
On my real hardware partition was much larger and no such problem appeared.
However, there was used "Windows 95 DOS", maybe that's why.
But this error deserves its own issue (if it's not a bug in game itself).

**Comment by vruppert** [comment]
Now the text should be displayed correctly.

**Comment by Vort** [comment]
Looks good now. Thank you.
![image](https://github.com/user-attachments/assets/e0694223-2525-4090-9cc0-52256bdc7619)

**Comment by fysnet** [comment]
> > Edit: The VGABIOS-elpin-2.40 BIOS does not suffer this issue and goes right to the 'No mouse driver' screen.
> 
> Do you use latest VGABIOS binaries? No more hangs for me.

Yes.
Using VGABIOS-lgpl-latest.bin, the VChess does and will crash because of the description I gave before.

Using VGABIOS-elpin-2.40 does not crash.
Using VGABIOS-lgpl-latest-cirrus.bin does not crash.
Using VGABIOS-lgpl-latest-banshee does not crash.

**Comment by Vort** [comment]
> Notice that the list of video screen numbers goes to 0x192 and then the 0xFFFF marker, for a total of 50 entries.

Ok, looks like I know what is going on here.
I have old monitor, so there are not many avaliable modes in this list.
Can you try setting `vbe_memsize=4` in config, will it help to lower mode count for you?
If not, there may be sense to allow downgrade of Bochs VBE further, to 2 MB of VRAM.

> This version of MS-DOS is assuming that there will not be more than 47 entries...

This code is from `VCHESS.EXE`:

![image](https://github.com/user-attachments/assets/d564c0e3-c959-42ee-98c3-9c1e5416d99d)

**Comment by fysnet** [comment]
That's the exact code I referenced, but I was showing it in the debugger, instead of a disassembly as you show.
I don't know of a way to tell the VGA BIOS to report less screen modes.
My question would be, don't you get the same results? If you use 
```
vgaromimage: file=/path/to/bios/VGABIOS-lgpl-latest.bin
```
don't you get the same results as I am reporting or does it boot to the game as expected?

**Comment by Vort** [comment]
> I don't know of a way to tell the VGA BIOS to report less screen modes.

As I said, You may try using `vbe_memsize=4` for it:

`vga: extension=vbe, update_freq=0, realtime=0, ddc=builtin, vbe_memsize=4`

_(In case you have [latest version](https://github.com/bochs-emu/VGABIOS/raw/20038d3ddef72ebb5f34c5dd71cdf25ebe9ad0b3/vgabios/VGABIOS-lgpl-latest.bin))_

> My question would be, don't you get the same results? If you use

Here is my mode list.
If I counted correctly, it have 38 entries instead of 50 in your case.
With such list game loads fine (except for mouse, but that is easy fix).

![image](https://github.com/user-attachments/assets/2d26d608-9cb8-403b-aa70-ff6b3853b182)

**Comment by fysnet** [comment]
Using the exact `bochsrc.bxrc` file you included with the original issue, adding
```
vga: extension=vbe, update_freq=0, realtime=0, ddc=builtin, vbe_memsize=4
```
and using the latest `VGABIOS-lgpl-latest.bin` file, I still get it to crash with a bunch of GPF messages.
I use a Windows build and it is up to date with the git.
The VGA BIOS has 50 entries for me.

**Comment by fysnet** [comment]
Here's my list, yours in read (identical to mine), then the few added in blue:
![mode_list](https://github.com/user-attachments/assets/07e63517-fd82-4d11-99a1-7547ab93449e)
I have no idea why mine has the extra added entries. Maybe Volker recognizes the mode numbers.
No big deal. I am not going to worry about it.
Thanks for your comments.

**Comment by Vort** [comment]
> (identical to mine)

No, not identical. My list is smaller.

> Maybe Volker recognizes the mode numbers.

It is easy to find them.
https://github.com/bochs-emu/VGABIOS/blob/20038d3ddef72ebb5f34c5dd71cdf25ebe9ad0b3/vgabios/vbe.c#L483-L490

Number 192h corresponds to 1920 x 1080 x 32.
The problem is 1920 * 1080 * 32 / 8 does not fit into 4 MB of VRAM.

What date is displayed when you start emulation?
![image](https://github.com/user-attachments/assets/5a5eaad7-0296-4bc5-895f-c900573399e5)

**Comment by Vort** [comment]
> Using the exact bochsrc.bxrc file you included with the original issue, adding

Similar line was already there, with `vbe_memsize=8`.
If you not just added another line, but edited existing one to `vbe_memsize=4`, then it means bug in VGABIOS.
I will try to reproduce it with VirtualBox, but slightly later.

**Comment by Vort** [comment]
I confirm problems with host resolution of 1920 x 1080 and `vbe_memsize=8` - game either hangs or prints errors in a loop.
But at the same time I observe normal operation with the same 1920 x 1080 in host and `vbe_memsize=4`.
Please check if your config have duplicated `vga: extension=vbe...` lines.
By the way, I think Bochs should warn user about duplicated lines.
I don't think such configuration "style" have any practical use.

**Comment by fysnet** [comment]
You're right. A ways further down in the `bochsrc` file, is a duplicate line with `vbe_memsize=8`. My fault for not checking, yet I figured if one was going to specify information about the VGA, it would be with the VGA BIOS declaration. :-)
When I removed the duplicate line, the `vbe_memsize=4` was now used and the errors are no longer there.

However, I do agree that maybe a check for duplicate lines should be added. Besides the fact that I failed to check myself, if there was a duplicate check, this would have been a bit easier to notice. Sorry about that.

**Comment by vruppert** [comment]
I have now added the memsize to the VBE init message. Since the latest VGABIOS checks the maximum supported resolution and the configured memory size I'm getting 40 VBE mode entries here with 16 MB, so there is no error.
Since I'm using a global bochsrc config file that is included by the specific ones, I don't like complaining about overriding settings. I noticed something like that in the USB code and I think it should be removed. Of course, if you save such a bochsrc with included global settings, it will be written to one file without duplicate lines.

**Comment by fysnet** [comment]
> Since I'm using a global bochsrc config file that is included by the specific ones, I don't like complaining about overriding settings. I noticed something like that in the USB code and I think it should be removed.

I see your point. However, couldn't we simply add a conditional build to it?

```
#if ALLOW_DUP_OVERRIDES == 0
  ... do a check to see if this config member has been initialized ...
#endif
```
This could even be built in to the config scripts so that you can build it with or without the checks.

**Comment by vruppert** [comment]
I think adding such checks for each config option would make things too complicated. Simply checking the bochsrc file for duplicate keywords would be bad since you can split the lines to make them better readable, e.g. configure each USB port in a separate line. Overriding bochsrc settings on the command line or change them in the start menu should work without complaining.
Useful checks could be added in bochsrc line parser to make sure all required parameters are given. In the USB case we could check if the port parameters are completely provided, e.g. if port1 should be set, we could check if options1 is also there (and vice versa).
